File: Source/Shelfinator.Runner/PaletteSequence.cpp
#include "PaletteSequence.h"

namespace Shelfinator
{
	namespace Runner
	{
		namespace PatternData
		{
			void PaletteSequence::Read(BufferFile::ptr file)
			{
				startTime = file->GetInt();
				endTime = file->GetInt();
				startPaletteIndex = file->GetInt();
				endPaletteIndex = file->GetInt();
			}

			double PaletteSequence::GetPercent(int time)
			{
				return startPaletteIndex == endPaletteIndex ? 0 : ((double)time - startTime) / (endTime - startTime);
			}
		}
	}
}

File: Source/Shelfinator.Creator/SongData/PaletteSequence.cs
using System.IO;

namespace Shelfinator.Creator.SongData
{
	class PaletteSequence
	{
		public int StartTime { get; set; }
		public int EndTime { get; set; }
		public int StartPaletteIndex { get; set; }
		public int EndPaletteIndex { get; set; }

		public PaletteSequence(int startTime, int endTime, int startPaletteIndex, int endPaletteIndex)
		{
			StartTime = startTime;
			EndTime = endTime;
			StartPaletteIndex = startPaletteIndex;
			EndPaletteIndex = endPaletteIndex;
		}

		public void Save(BinaryWriter output)
		{
			output.Write(StartTime);
			output.Write(EndTime);
			output.Write(StartPaletteIndex);
			output.Write(EndPaletteIndex);
		}

		public override string ToString() => $"Time: {StartTime}-{EndTime}, Index: {StartPaletteIndex}-{EndPaletteIndex}";
	}
}
